### PR TITLE
feat: walletconnect v2 wallet

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

Adds the ability to connect wallets via WalletConnectV2 on the following EVMs (simultaneously):

- Ethereum
- Polygon
- Optimism
- Gnosis
- Avalanche
- Binance

Note, this PR only enables it in `develop` for now so ops can give it a good test and for it to not block release if anything needs attention.

I plan to address any non-blocking comments/nip-picks in a followup PR if necessary before going live.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Closes https://github.com/shapeshift/web/issues/5228

## Risk

Medium:

- Removes existing (non-working) WalletConnectV1 functionality
- Adds new WalletConnectV2 wallet connection functionality that could cause issues if signing was implemented incorrectly

## Testing

Connect to a wallet via WalletConnectV2 (ideally one that supports multiple chains, e.g. Trust wallet).

Screenshot 2023-09-25 at 3 13 57 pm

Confirm that modal displays, and upon connecting via the wallet, the ShapeShift app redirects to the Dashboard with the account connected.

Confirm that transactions can be signed (check multiple chains and transaction types).

### Engineering

☝️

### Operations

☝️

## Screenshots (if applicable)

image

image

### CODE REVIEW

**Comment by 0xApotheosis** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #5315** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5315" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/5315?utm_source=stack-comment).

File: src/plugins/walletConnectToDapps/utils/clearAllWalletConnectToDappsSessions.ts [hunk 1]
```diff
@@ -25,7 +27,4 @@ export const clearAllWalletConnectToDappsSessions = async () => {
       reason: getSdkError('USER_DISCONNECTED'),
     })
   }
-
-  // catch-all, clear local storage
-  clearWalletConnectLocalStorage()
```
Comment (0xApotheosis) [inline src/plugins/walletConnectToDapps/utils/clearAllWalletConnectToDappsSessions.ts:30]: This was rugging the WC wallet state, and is only here as a catch-all to clear wallet-connect-to-dApp sessions, and shouldn't really be needed.

File: .env.base [hunk 2]
```diff
@@ -152,5 +152,10 @@ REACT_APP_SNAP_ID=npm:@shapeshiftoss/metamask-snaps
 
 REACT_APP_EXPERIMENTAL_MM_SNAPPY_FINGERS=true
 
+REACT_APP_SNAP_ID=npm:@shapeshiftoss/metamask-snaps
```
Comment (0xApotheosis) [inline .env.base:155]: TODO: what's up with this diff?

Comment (gomesalexandre) [reply .env.base]: Interactively dropped the additional commits that were present here

File: src/context/WalletProvider/WalletConnectV2/useWalletConnectV2EventHandler.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    /**
+     * in case of KeepKey placeholder wallet,
+     * the disconnect function is undefined
+     */
+    state.wallet?.disconnect?.()
+    dispatch({ type: WalletActions.RESET_STATE })
+    clearLocalWallet()
+  }, [dispatch, state.wallet])
+
+  useEffect(() => {
```
Comment (gomesalexandre) [inline src/context/WalletProvider/WalletConnectV2/useWalletConnectV2EventHandler.ts:26]: Can we add the same guards as this guy? 🙏🏽 

Event handlers running on any keyring (or in this case, any Ethereum provider) are introducing a bunch of bugs

https://github.com/shapeshift/web/pull/5336/files#diff-850c3685f1d3ca0096c3204a6bbf59b35b1780e11e57057d75eec034be4206b9R33-R37

File: src/context/WalletProvider/WalletConnectV2/config.ts [hunk 4]
```diff
@@ -1,11 +1,29 @@
...
   icon: WalletConnectIcon,
-  name: 'WalletConnect',
+  name: 'WalletConnectV2',
   description: 'v2',
 }
+
+export const walletConnectV2ProviderConfig: EthereumProviderOptions = {
+  projectId: getConfig().REACT_APP_WALLET_CONNECT_PROJECT_ID,
+  chains: [1],
+  optionalChains: [10, 56, 100, 137, 43114],
```
Comment (gomesalexandre) [inline src/context/WalletProvider/WalletConnectV2/config.ts:18]: Nitpick, can we either make this an IIFE and use asset data, or at least comment which chains these Ethereum network IDs refer to?

File: src/context/WalletProvider/WalletConnectV2/Web3ModalProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+type Web3ModalContextType = {
+  ethereumClient: EthereumClient
+}
+
+export const Web3ModalContext = createContext<Web3ModalContextType | undefined>(undefined)
+
+// Initialization functions
+const createEthereumClient = () => {
+  const chains = [mainnet]
```
Comment (gomesalexandre) [inline src/context/WalletProvider/WalletConnectV2/Web3ModalProvider.tsx:21]: Do we want to match `optionalChains` here?

**Review by gomesalexandre — COMMENTED** [review]
Tested locally, LGTM overall - found one bug WRT CoW message signing, and a few nitpicks:

### Connection

image

We may want some @reallybeard oil here to reduce the font size for this specific wallet option and have more spacing between the `WalletConnectV2` option and the icon - also, this should probably be `WalletConnect V2`, or even `WalletConnect` as users shouldn't be concerned with the V1/V2 implementation details and V2 is the de-facto one now? Note in the following screenshot, this is referred to as `WalletConnect` 

image
image

image

- Same here, should probably be `WalletConnect`

### Sends

image
https://gnosisscan.io/tx/0xb8b8b3f6e3b81d3d0000938e480838826f96f3ce10f449aca242960849808a4b

### Receives

image

### Swaps

#### CoW message signing


image

- This never pops a signature request (tested with Trust wallet)

### Li.Fi (regular Tx)

image
https://gnosisscan.io/tx/0x8c776a50c0aa7d20e314346650dd9ce9881ebbc890c981ee12c9ec5336fabc69

File: src/context/WalletProvider/WalletConnectV2/useWalletConnectV2EventHandler.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  dispatch: Dispatch<ActionTypes>,
+) => {
+  const { provider } = state
+  const ethereumProvider = provider as EthereumProvider
+
+  const handleDisconnect = useCallback(() => {
+    /**
+     * in case of KeepKey placeholder wallet,
+     * the disconnect function is undefined
+     */
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/context/WalletProvider/WalletConnectV2/useWalletConnectV2EventHandler.ts:20]: I.. can't see a difference here? 😅

File: src/context/WalletProvider/WalletConnectV2/components/Connect.tsx [hunk 7]
```diff
@@ -1,99 +1,73 @@
...
-  // The WalletConnect modal handles desktop and mobile detection as well as deep linking
   return (
     <ConnectModal
       headerText={'walletProvider.walletConnect.connect.header'}
       bodyText={'walletProvider.walletConnect.connect.body'}
       buttonText={'walletProvider.walletConnect.connect.button'}
       onPairDeviceClick={pairDevice}
       loading={loading}
-      error={error}
+      error={null}
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/context/WalletProvider/WalletConnectV2/components/Connect.tsx:70]: Made things nicer in https://github.com/shapeshift/web/pull/5315/commits/d2465f79fe6df014a29782a9277cf85c7fe91dfc

File: src/context/WalletProvider/WalletConnectV2/Web3ModalProvider.tsx (hunk 5 above)
Comment (0xApotheosis) [reply src/context/WalletProvider/WalletConnectV2/Web3ModalProvider.tsx]: We do! Addressed in https://github.com/shapeshift/web/pull/5315/commits/bae3cf8f51be00329e9ead4e297b3eca523be7b3.

File: src/context/WalletProvider/WalletConnectV2/config.ts (hunk 4 above)
Comment (0xApotheosis) [reply src/context/WalletProvider/WalletConnectV2/config.ts]: Fixed this logic up in https://github.com/shapeshift/web/pull/5315/commits/bae3cf8f51be00329e9ead4e297b3eca523be7b3 to use the Viem chains.

File: src/context/WalletProvider/WalletConnectV2/useWalletConnectV2EventHandler.ts (hunk 3 above)
Comment (0xApotheosis) [reply src/context/WalletProvider/WalletConnectV2/useWalletConnectV2EventHandler.ts:27]: 🫡 https://github.com/shapeshift/web/pull/5315/commits/f8095de7c2fa7433f7a2c6493e0c1e23c625c0d0

File: src/context/WalletProvider/WalletConnectV2/components/Connect.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/context/WalletProvider/WalletConnectV2/components/Connect.tsx:70]: We still need it, as it's a shared component.

**Comment by 0xApotheosis** [comment]
@gomesalexandre all comments addressed except the CoW issue, which I'll fix in a follow-up PR.

**Review by woodenfurniture — APPROVED** [review]
approving since this is behind feature flag and cowswap fixes will be raised in follow-up pr before enabling

**Pull request merged**